### ISSUE TITLE
[YouTube[ Import Playlist

### ISSUE THREAD

**Issue by temportalflux** [temportalflux/MusicAssist#4]
Add support for importing entire youtube playlists

**Comment by temportalflux** [comment]
Unrelated, but I highly recommend using ZenHub. Its how ticket progress is being tracked right now.

**Comment by casualchameleon** [comment]
Sounds good - do I need an invite to that or shall I handle my own workspace?

**Comment by casualchameleon** [comment]
I agree with your approach to using minimal patching. How about a button on the playlist creation dialog that opens a completely new configuration dialog that can be used to handle youtube playlist-specific functions?
I'm really not a fan of having a single track represent an entire playlist for the reasons you've mentioned. I was thinking more along the lines of having a foundry playlist represent a youtube playlist. I think some kind of routine that periodically fire off a youtube data api request to check each url is still reachable. This can be done all at once in a single call using the [list](https://developers.google.com/youtube/v3/docs/videos/list) endpoint so it's not too chatty. We could maybe highlight broken playlist videos in red and skip them in the playlist?

**Comment by casualchameleon** [comment]
I've been doing some reading and you need an API key to access the youtube data api. Fortunately, it's completely free, and just has a daily quota of 10,000. 
We can't really include one with the module as we'd expose our key to quota theft and other abuse.
So, I propose we create a setting in the module settings for the user to add this API key to share with other users on their table. 

What do you think?

**Comment by temportalflux** [comment]
Hmm lots of problems to solve. I curious what the next version of foundry (with updated audio support being announced) will bring. I'll update this ticket if/when PRs get submitted, reviewed, and committed, so no need to make a second workspace unless you want one for your fork.
1. Data API Key - That IS a pickle. I suspect the indented functionality is for a Client Server framework where the server is the only owner of the key. If this module is ever integrated with core FVTT, we may be able to use that key. I am not a fan at all of requiring users to go create a developer application and get involved in the nitty gritty of youtube api stuff. It also breaks the plugin-and-play methodology that all FVTT Modules have.
3. Playlists - I'm not entirely sure how the UX for playlist management should be implemented. I had some success in adding a button to the sidebar for creating playlists/tracks with a new modal, though later removed that code when I favored patches to pre-existing modals. You could add a new button for synced-youtube-playlists which adds a flag to the playlist object denoting that the playlist is `streamed` (like tracks do). On init/ready Hook (whenever FVTT is opened or refreshed), the playlists can be scanned for any which have the `streamed` flag. If one does (and we can resolve some way of determining what videos are in a youtube playlist), then an asynchronous request can be made to youtube, and when it returns, we diff against the known tracks in the playlist. During the diff, remove any no longer in the youtube playlist, add any new ones to the FVTT playlist as a streamed track. This would also mean the user cannot edit tracks in `streamed` playlists.

**Comment by casualchameleon** [comment]
Yeah, I think the data API key is a dead-end for now. I agree that getting users to create their own keys is a bit of a PITA and not a pleasant experience. I think, like you said, we'll need to wait to see what the folks at foundry bring in the future.
As for your second point, I think I'll take a whack at implementing this and see how far I get. I'll need to make some changes to the build/deploy process as well, as at the moment there is a hard-coded ref pointing at your PC's documents folder.
Have you used gulp? I am fairly confident I could get some really nice deployment tools running with that. I may do this anyway in my fork, because as it stands the python script just crashes when I run it! We can always cherrypick later.

**Comment by temportalflux** [comment]
HA So there is! I forgot about that! I think if you just change it locally its fine. Thats maybe 1 line of the module.json which should honestly not exist there because it is only for development purposes. Maybe I should add another json file for build settings...

**Comment by casualchameleon** [comment]
Yeah, other projects I've looked at have set up a foundryconfig.json file that is -not- checked in to source control. That way it is private and you don't leak any dev-specific details into the repo. For now though I have just changed the ref in module.json. I just won't check in that change!

**Comment by casualchameleon** [comment]
I've made good headway into this. 
I have an 'Import Youtube Playlist' button working, which opens a custom formapp
![image](https://user-images.githubusercontent.com/1485053/85184290-22a83700-b287-11ea-89d2-d4b3042d22c5.png)

![image](https://user-images.githubusercontent.com/1485053/85184273-0b694980-b287-11ea-897a-3cde7bb7307b.png)

I've run into a snag though. The Youtube IFrame API will only return Ids, so we have no way of getting track names. player.getPlaylist() will give us all the video Ids, so we can add them all to a playlist no problem, but there's not a clean way to go and get all of the track names (without the data API!)

There is, however, an unclean way! Within the generated iframe for a playlist is a list of queued tracks
![image](https://user-images.githubusercontent.com/1485053/85185119-8c761000-b28a-11ea-9333-43555c0290aa.png)

Within this are two key items. The video Id, and the video title.
![image](https://user-images.githubusercontent.com/1485053/85185155-baf3eb00-b28a-11ea-8fbc-e6c901d96348.png)

I think these are trivial to scrape, and using these we can have a streamed track video Id and the name auto-generated for the user. I'm going to give this a try this weekend.

**Comment by temportalflux** [comment]
Haha this sounds about right for a module hacking into the youtube iframe api! I have my doubts about being able to access the inner html, but it's nice to see the progress. Most of my side projects are on hold atm due to irl work, but nice headway!

**Comment by casualchameleon** [comment]
borrowed (stolen) from [stack overflow](https://stackoverflow.com/questions/1452871/how-can-i-access-iframe-elements-with-javascript)

function iframeRef( frameRef ) {
    return frameRef.contentWindow
        ? frameRef.contentWindow.document
        : frameRef.contentDocument
}

where iframeRef is the iframeId.

**Comment by casualchameleon** [comment]
Alright, in my sleep-addled mind yesterday I fell victim to forgetting about cross-origin policy... woops. Iframe scraping isn't going to work. I think I'm going to have to iterate over each video ID and load it in the player, and grab the player.getVideoData().title for each one. Not ideal - this could get pretty slow for big playlists. I'll do some testing!

**Comment by casualchameleon** [comment]
Got this working. I just need to hook up the UI and then create the sounds based off the vidIds and track names. Changes are sitting in my fork in the playlist-func branch. It's not too bad. Takes a couple of seconds to do ~100 tracks but it's all async so won't hang the UI. Hammers network though!

**Comment by casualchameleon** [comment]
Finally got round to finishing this off. Ready to go!
